pa1/PPCodeUnitStream.h

```diff
@@ -0,0 +1,55 @@
+#ifndef PPCodeUnitStream_h
+#define PPCodeUnitStream_h
+
+#include "PPCodeUnit.h"
+#include "UTF32StreamIfc.h"
+#include "PPCodeUnitStreamIfc.h"
+#include <queue>
+#include <string>
+
+// Stream UTF32 code points into stream of PPCodeUnit that is easy to tokenize.
+//
+//   - Line splicing. i.e., occurrences of "\\\n" are replaced with
+//     PPCodeUnitType::WhitespaceCharacter so that we can revert it in raw strings.
+//
+//   - Comments are streamed into PPCodeUnitType::Comment objects. Other tools
+//     can extract the comment strings for more processing
+//
+//   - Digraphs and universal-character-names are also extracted out and can be
+//     reverted in raw strings.
+class PPCodeUnitStream: public PPCodeUnitStreamIfc {
+public:
+  PPCodeUnitStream(std::shared_ptr<UTF32StreamIfc>);
+
+  virtual bool isEmpty() const override;
+  virtual const std::shared_ptr<PPCodeUnit> getCodeUnit() const override;
+  virtual void toNext() override;
+
+  std::string getErrorMessage() const;
+
+private:
+  std::shared_ptr<UTF32StreamIfc> _u32stream;
+
+  // Error reporting related stuff.
+  void _setError(const std::string&&);
+  void _clearError();
+  std::string _errorMessage;
+
+  // _pushCodeUnits() parses the input stream and pushes one or more PPCodeUnits
```

Comment:
Maybe I should document the reason for not allowing backward movement of the stream somewhere in the source file:

In summary, no need for it.

Though the instinct are probably that allowing the users of these stream classes, including `UTF32StreamIfc`, `PPCodeUnitStreamIfc`, and `PPTokenizerDFA`, are easier to use with backward movement functions such as `moveOneUnitBackward()` or similar. In reality, the entire preprocessing tokenizer can be constructed without them just fine.

That said, though, it is a long time debate and exploration that I finally found this current interface most suitable. You can move forward by one unit with `toNext()`, get the current unit by `getCodeUnit()`, and check for error messages with `getErrorMessage()`. Note how this same API pattern propagates through `UTF32StreamIfc`, `PPCodeUnitStreamIfc`, and `PPTokenizerDFA` unchanged. As of this moment, I tend to believe that these are the minimal API for getting the work done lazily - that is - done as late as possible through these layers.